### PR TITLE
LibJS: Make test262 run on serenity

### PR DESCRIPTION
**Pull request by davidot**
The runner should replace the runner used in linusg/libjs-test262 eventually, however this will require some work on the main.py script there.

If you are interested in the diff between inside serenity vs linux here it is:
[serenity-diff2.txt](https://github.com/SerenityOS/serenity/files/9540691/serenity-diff2.txt)

Also I had to add some emojis which we use so not sure if I did everything right there.

Results:
test262-full3

(Note has old icon for :gear: gear and my icon instead of the master one for :boom: collision/explosion)

### CODE REVIEW

File: Tests/LibJS/test262-runner.cpp [hunk 1]
```diff
@@ -0,0 +1,858 @@
+/*
+ * Copyright (c) 2021, Linus Groh <linusg@serenityos.org>
+ * Copyright (c) 2021-2022, David Tuin <davidot@serenityos.org>
```
Comment (davidot) [inline Tests/LibJS/test262-runner.cpp:3]: Is just keeping the copyright like this fine? I started by copying the file

**Comment by linusg** [comment]
Strongly against maintaining two separate versions separately, let's first upstream main.cpp and then add serenity support to it

**Comment by davidot** [comment]
Split the commit importing main.cpp -> test262-runner.cpp and fixing it up and putting it on Serenity.

File: Tests/LibJS/test262-runner.cpp [hunk 2]
```diff
@@ -502,20 +502,39 @@ static bool verify_test(Result<void, TestError>& result, TestMetadata const& met
 }
 
 static FILE* saved_stdout_fd;
+static bool in_assert = false;
```
Comment (linusg) [inline Tests/LibJS/test262-runner.cpp:526]: `g_in_assert`

File: Tests/LibJS/test262-runner.cpp [hunk 3]
```diff
@@ -172,11 +172,36 @@ static Result<JS::NonnullGCPtr<JS::Script>, TestError> parse_harness_files(JS::R
...
+    StringView type;
+};
+
+static Result<void, TestError> run_test(StringView source, StringView filepath, TestMetadata const& metadata)
 {
-    if (s_parse_only) {
+    if (s_parse_only || (metadata.is_negative && metadata.phase == NegativePhase::ParseOrEarly && metadata.program_type != JS::Program::Type::Module)) {
         // Creating the vm and interpreter is heavy so we just parse directly here.
-        auto parser = JS::Parser(JS::Lexer(source, filepath), program_type);
+        // We can also skip if if we now the test is supposed to fail during parse
```
Comment (linusg) [inline Tests/LibJS/test262-runner.cpp:201]: "skip if if"

**Review by linusg — COMMENTED** [review]
Great work!

File: Tests/LibJS/test262-runner.cpp [hunk 4]
```diff
@@ -0,0 +1,791 @@
...
+    bool is_negative { false };
+    NegativePhase phase { NegativePhase::ParseOrEarly };
+    StringView type;
+};
+
+static Result<void, TestError> run_test(StringView source, StringView filepath, TestMetadata const& metadata)
+{
+    if (s_parse_only || (metadata.is_negative && metadata.phase == NegativePhase::ParseOrEarly && metadata.program_type != JS::Program::Type::Module)) {
+        // Creating the vm and interpreter is heavy so we just parse directly here.
+        // We can also skip if we now the test is supposed to fail during parse
```
Comment (linusg) [inline Tests/LibJS/test262-runner.cpp:201]:
```suggestion
        // We can also skip if we know the test is supposed to fail during parse
```

oops, one more

Comment (davidot) [reply Tests/LibJS/test262-runner.cpp]: I don't now why I keep making this mistake know

**Comment by davidot** [comment]
All my emoji got merged already :upside_down_face: so I've removed that commit and @Xexxa can merge the gear one themself.

File: .github/workflows/libjs-test262.yml [hunk 5]
```diff
@@ -95,13 +95,13 @@ jobs:
           path: ${{ github.workspace }}/libjs-test262/Build/CLDR
           key: UnicodeData-${{ hashFiles('Meta/CMake/locale_data.cmake') }}
 
-      - name: Build libjs-test262-runner and test-js
+      - name: Build libjs-test262, test-js and test-wasm
```
Comment (linusg) [inline .github/workflows/libjs-test262.yml:98]:
```suggestion
      - name: Build test262-runner, test-js and test-wasm
```

Comment (davidot) [reply .github/workflows/libjs-test262.yml]: Whoops, chopped the wrong word

**Pull request merged**